opentelemetry-jaeger/src/exporter/config/mod.rs

```diff
@@ -0,0 +1,268 @@
+//!
+//! [jaeger agent]: https://www.jaegertracing.io/docs/1.31/deployment/#agent
+//! [jaeger collector]: https://www.jaegertracing.io/docs/1.31/deployment/#collector
+//! [jaeger deployment guide]: https://www.jaegertracing.io/docs/1.31/deployment
+
+use crate::Process;
+use opentelemetry::sdk::trace::Config;
+use opentelemetry::sdk::Resource;
+use opentelemetry::trace::{TraceError, TracerProvider};
+use opentelemetry::{global, sdk, KeyValue};
+use opentelemetry_semantic_conventions as semcov;
+use std::sync::Arc;
+
+/// Config a exporter that sends the spans to a [jaeger agent](https://www.jaegertracing.io/docs/1.31/deployment/#agent).
+pub mod agent;
+/// Config a exporter that bypass the agent and send spans directly to [jaeger collector](https://www.jaegertracing.io/docs/1.31/deployment/#collector).
+#[cfg(any(feature = "collector_client", feature = "wasm_collector_client"))]
+pub mod collector;
+
+mod common {
+    use opentelemetry::sdk;
+
+    // configurations and overrides on how to transform OTLP spans to Jaeger spans.
+    #[derive(Debug)]
+    pub struct TransformationConfig {
+        pub export_instrument_library: bool,
+        pub service_name: Option<String>,
+    }
+
+    impl Default for TransformationConfig {
+        fn default() -> Self {
+            TransformationConfig {
+                export_instrument_library: true,
+                service_name: None,
+            }
+        }
+    }
+
+    // pipeline must have transformation config and trace config.
+    pub trait HasRequiredConfig {
+        fn set_transformation_config<T>(&mut self, f: T)
+        where
+            T: FnOnce(&mut TransformationConfig);
+
+        fn set_trace_config(&mut self, config: sdk::trace::Config);
+    }
+}
+
+/// Common configurations shared by the agent and the collector. You can use those functions
+/// in [`CollectorPipeline`] and [`AgentPipeline`].
+/// Generally there are three types of configurations:
+///
+/// - configurations required by Jaeger. For example, `service_name`.
+/// - performance configurations for the jaeger exporter. For example, `instrumentation_library_tags`
+/// - opentelemetry trace configurations.
+///
+/// # Example
+/// ```
+/// use opentelemetry_jaeger::{Configurable, new_agent_pipeline};
+/// use opentelemetry::{sdk, sdk::Resource, KeyValue};
+///
+/// let pipeline = new_agent_pipeline()
+///                 .with_service_name("test-service")
+///                 .with_trace_config(
+///                       sdk::trace::Config::default()
+///                         .with_resource(Resource::new(
+///                             vec![KeyValue::new("service.name", "my-service")])
+///                         )
+///                 );
+/// ```
+///
+/// [`CollectorPipeline`]: crate::config::collector::CollectorPipeline
+/// [`AgentPipeline`]: crate::config::agent::AgentPipeline
+pub trait Configurable: Sized {
```

Comment:
Hm is there a way to not require the end user to use this trait?

Reply:
We can delete this trait and just copy paste the method into the agent pipeline and collector pipeline? We should probably still keep the `HasRequiredConfig` trait to make sure the two pipeline has the common configurations we want.

Reply:
Yeah copying the methods probably has enough of a developer experience benefit that it's worth the maintenance cost